Step5 - 자동차 경주(리팩토링)

gmltkddl1 opened the pull request:
이전 PR에서 주신 피드백을 반영했고
피드백 주신대로 요구사항에 맞춰 commit을 생성하였습니다.
2번째 요구사항인 domain이 view를 의존하지 않는다는 이미 구현되어있어 생략되었습니다.

gmltkddl1 committed:
refactor: step4 자동차 경주(우승자) 피드백 반영

gmltkddl1 committed:
feat: step5 자동차 경주(리팩토링)

핵심 비지니스 로직을 가지는 객체를 domain 패키지, UI 관련한 객체를 view 패키지에 구현한다.

gmltkddl1 committed:
feat: step5 자동차 경주(리팩토링)

테스트 가능한 부분과 테스트하기 힘든 부분을 분리해 테스트 가능한 부분에 대해서만 단위 테스트를 진행한다.

Flamme1004K commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -1,5 +1,9 @@
+import racingcar.domain.Race
+import racingcar.view.InputView
+import racingcar.view.OutputView
+
```
```
 for (i in 1..numberOfTrys) {
        race.process()
        outputView.drawRace(race)
    }
```

레이싱을 n번 수행한다는 부분은 도메인 로직일까요? 아니면, ui와 도메인 로직이 섞인 로직일까요?

gmltkddl1 replied:
ui와 도메인이 섞인 로직으로 보입니다.
그런데 요구사항을 보면 race의 진행사항을 보여줘야하고 결국 코드의 특정 부분에서는 ui와 도메인을 연결해줘야 생각하는데
이 부분에도 개선할 점이 있을까요?

gmltkddl1 replied:
최대한 ui와 domain을 분리하는 리팩토링을 진행해보았습니다.

Flamme1004K replied:
잘해주셨어요! 단, 현재의 도메인 로직이 테스트에 용이한지 고민해보시길 바래요!

Flamme1004K commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -1,19 +1,18 @@
+    var moveCount: Int = initialMoveCount
+        private set(value) {
+            field = value
+        }
```
```suggestion
        private set
```

이렇게 써주셔도 됩니다~

Flamme1004K commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -1,19 +1,18 @@
 
     init {
-        if (name.length > 5) throw IllegalArgumentException("자동차 이름이 너무 길어요")
+        require(name.length <= 5) { "자동차 이름이 너무 길어요" }
```
5에 대해서도 상수로 빼볼까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -1,19 +1,18 @@
-
     fun process() {
         val randomNumber = RandomNumberGenerator().getRandomNumber()
         move(randomNumber)
```
<img width="975" alt="스크린샷 2024-11-19 오후 7 38 01" src="https://github.com/user-attachments/assets/ccf57b36-440f-4f35-af0d-6e98636368fe">

요구 사항에 위와 같은 그림이 있을거에요. 과연 지금의 자동차가 이동에 대한 단위 테스트가 가능한지 살펴보고, 테스트 가능한 자동차로 만들어보시겠어요?

Flamme1004K commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -1,29 +1,31 @@
+        sb.append(
+            race.getWinner().map { it.name }
+                .joinToString(","),
+        )
```
joinToString을 사용하면 StringBuilder가 필요 없겠죠?

Flamme1004K commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -1,29 +1,31 @@
         sb.append(car.name)
         sb.append(" : ")
-        for (i in 1..car.status()) {
+        for (i in 1..car.moveCount) {
```
이것도 repeat로 변경 할 수 있어보이네요~

Flamme1004K commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -1,19 +1,18 @@
 
 class Car(
     initialMoveCount: Int = 0,
     val name: String = "Car",
```
name에 대해서 "Car" 를 default 값으로 넣은 이유를 알 수 있을까요?

gmltkddl1 replied:
테스트 작성시에 편리성을 위해서 만들었는데 다시 생각해보니 요구사항에는 불필요한 부분 같아 삭제하겠습니다

Flamme1004K commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -2,14 +2,15 @@ package racingcar
         assertThat(race.cars.size).isEqualTo(3)
-        race.cars.forEach { assertThat(it.status()).isEqualTo(0) }
-        race.cars.forEachIndexed { index, car -> assertThat(car.name).isEqualTo("car${index + 1}") }
+        race.cars.forEach { assertThat(it.moveCount).isEqualTo(0) }
```
foreach를 돌리는 것보다, containsExactly 을 사용하여 리스트 비교가 더 좋지 않을까하네요~

Flamme1004K commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -19,7 +20,6 @@ class RaceTest {
         race.cars[1].move(4)
```
이제 배열에서 꺼내서 자동차를 움직여 우승자를 만들지 않고, 주어진 횟수 동안 자동차를 움직이는 메서드를 만들고 우승자를 만들어보세요.

Flamme1004K commented on README.md:
```diff
@@ -1 +1,16 @@
-# kotlin-racingcar
\ No newline at end of file
+# kotlin-racingcar
+
+기능 요구사항
+
```
기능 요구 사항 정리 잘하셨어요!

Flamme1004K reviewed (changes requested):
안녕하세요! 

5단계 미션 구현 잘 해주셨어요!  몇 가지 코멘트 남겼으니 적용 후 다시 리뷰 요청 부탁드립니다.

감사합니다!

gmltkddl1 committed:
refactor: step5 피드백 반영

private set 생략 가능한 부분 삭제
매직넘버 상수로 변경
RandomNumberGenerator를 car가 아닌 race에서 사용하여 단위테스트 가능하도록 변경
불필요한 stringbuilder 삭제
for문을 repeat로 변경
car클래스의 name default value 삭제
List에 대한 결과 검증시 반복문 대신 containExactly사용

gmltkddl1 committed:
refactor: step5 피드백 반영

List에 대한 결과 검증시 반복문 대신 containExactly사용

gmltkddl1 committed:
refactor: step5 피드백 View 로직과 Domain 로직을 분리

Flamme1004K commented on src/main/kotlin/racingcar/domain/Race.kt:
```diff
@@ -0,0 +1,12 @@
+package racingcar.domain
+
+class Race(
+    names: List<String>,
+) {
+    val cars: List<Car> = names.map { Car(name = it) }
```
```suggestion
class Race(
    val cars: List<Car>
) {
```

초기화 파라미터를 사용하여 프로퍼티를 만드는 부분과, 프로퍼티를 바로 주입하는 부분의 차이는 객체 생성 관점에서 어떠한 장단점이 있을까요?

gmltkddl1 replied:
프로퍼티에 바로 주입하면 코드가 간단해지는 장점이 있을거같고
초기화 파라미터로 초기화하면 객체 생성을 제가 원하는 방식으로만 생성을 제약할수 있을거 같습니다.

프로퍼티에서 초기화하면 시작부터 특정 차량의 이동수를 0이 아닌 다른 값으로 줄 수 있고 
이것은 객체를 유연하게 사용할수 있게 할 수 있겠지만
현재 자동차 경주는 출발점이 무조건 0이므로 위와 같이 생성 방식을 제한하는 것이 더 좋다고 생각했습니다.

Flamme1004K replied:
매우 잘 파악해주셨어요! 

저는 프로퍼티에서 초기화하면 테스트에 용이한 생성자를 만들 수가 있다고 생각해요. 테스트를 위한 생성자는 너무 테스트를 위해서 풀어놓는게 아닌가요? 하는 의견도 있겠지만, 테스트에 용이하게 생성자에 대한 룰을 풀어놓는 부분은 같이 일하는 동료 간의 약속으로 풀어나갈 수 있어보여요.

만약에 자동차에 대한 생성에 대해 꼭 name을 써야되겠다 하면, [정적 팩토리 메서드 패턴](https://velog.io/@lychee/%EC%9D%B4%ED%8E%99%ED%8B%B0%EB%B8%8C-%EC%9E%90%EB%B0%94-%EC%95%84%EC%9D%B4%ED%85%9C-1.-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%84%B0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C%EB%A5%BC-%EA%B3%A0%EB%A0%A4%ED%95%98%EB%9D%BC)을 통하여 from(), of() 같은 정적 메서드를 만들어 정적 메서드 내에서 자동차를 만들고 Race를 만들어 볼 듯 하네요.

한번 정적 팩토리 메서드로 자동차 생성에 대한 부분을 개선해 볼까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/Race.kt:
```diff
@@ -0,0 +1,12 @@
+    val cars: List<Car> = names.map { Car(name = it) }
+
+    fun doRace(numbers: List<Int>) {
+        require(cars.size == numbers.size)
```
예외에 대한 메시지는 필요하지 않을까요~

Flamme1004K commented on src/main/kotlin/racingcar/domain/Race.kt:
```diff
@@ -0,0 +1,12 @@
+
+    fun doRace(numbers: List<Int>) {
+        require(cars.size == numbers.size)
+        repeat(cars.size) { index -> cars[index].move(numbers[index]) }
```
index를 이용하여 자동차를 움직여야하는 이유를 알 수 있을까요?

gmltkddl1 replied:
구현 방식을 doRace에 각 차의 해당되는 랜덤 넘버를 넘겨서 레이스를 진행하는 방식으로 구현하였습니다.
예를들어 
race(listof("car1","car2","car3"))
doRace(listof(4,5,1))
car1의 move에 4
car2의 move에 5
car3의 move에 1을 넘겨줘야하여 위와 같이 구현했습니다.

gmltkddl1 replied:
차 한대별로 랜덤 넘버를 시행 횟수 만큼 받아서
레이스를 진행하고 결과를 출력하는 식으로 바꾸면 index를 제거할수있을것으로 보이긴합니다만
위 구현이 크게 문제가 될까요?

Flamme1004K replied:
오히려 이상하다고 생각해요. 

지금 코드 상의 흐름은 레이스가 시작되기 전에 이미 누가 승리하는지에 대한 결과가 정해져있어요. 

```
val numbersLists: MutableList<List<Int>> = mutableListOf()
    repeat(numberOfTrys) {
        numbersLists.add(randomNumberGenerator.getRandomNumbers(names.size))
    }
```

바로 main 함수에 있는 코드로 인해서죠. 

저는 이렇게 랜덤 값을 미리 만드는 부분보다, 자동차가 move() 할 때 랜덤 값이 그때그때 결정되었으면 해요.

Flamme1004K replied:
이 코드가 개선되는 방향에 대해서는 나중에 많은 도움이 될거에요. 일단 랜덤 값을 미리 만드는 부분보다, 함수를 넣어서 실행시키는 방향부터 진행해보는 건 어떨까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceGame.kt:
```diff
@@ -0,0 +1,13 @@
+            race.doRace(randomNumberGenerator.getRandomNumbers(race.cars.size))
+            raceRecords.raceRecords.add(RaceRecord(race = race))
+        }
+    }
```
여기서 RandomNumberGenerator를 사용하여 자동차를 움직인다면, RaceGame를 통하여 우승자를 어떻게 만들 수 있을까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceGame.kt:
```diff
@@ -0,0 +1,13 @@
+    fun doRaceGame(numberOfTry: Int) {
+        repeat(numberOfTry) {
+            race.doRace(randomNumberGenerator.getRandomNumbers(race.cars.size))
+            raceRecords.raceRecords.add(RaceRecord(race = race))
```
```suggestion
            raceRecords.add(RaceRecord(race))
```

raceRecords에서 raceRecords를 이용하여 RaceRecord를 더한다는 부분보다, raceRecords에 add 함수를 만들어서 raceRecords 프로퍼티를 감추어 캡슐화해 주는 건 어떨까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceRecords.kt:
```diff
@@ -0,0 +1,15 @@
+package racingcar.domain
+
+class RaceRecords {
+    val raceRecords: MutableList<RaceRecord> = mutableListOf()
```
불변으로 만들 수 있지 않을까요?

gmltkddl1 replied:
add가 필요한 리스트인데 불변으로 바꿀수 있나요?

Flamme1004K replied:
넵! 가능합니다. 

```
fun doRaceGame(numberLists: List<List<Int>>) {
        numberLists.forEach {
            race.doRace(it)
            raceRecords.addRaceRecord(RaceRecord(race = race))
        }
    }
```

RaceGame 클래스의 numberLists.forEach를 map으로 변경하면 어느정도 감이 올거에요!

gmltkddl1 replied:
제가 이해가 잘 안되는데 위 리스트는 실제로 추가가 필요한 리스트이고 피드백 주신대로 map으로 새로운 리스트를 만들게 되면
var raceRecords: List<RaceRecord> = listOf() 이런식이 되야 할거 같습니다.
하지만 var 역시 가변이라고 생각되거든요
val raceRecords: MutableList<RaceRecord> = mutableListOf() 이 구조보다 위와 같은 구조가
더 좋은 것 일까요? 
아니면 그때 그때 아예 raceRecords를 새로 생성해는데 그렇게해도 이 객체를 프로퍼티로 가지는 객체가 가변인자를 갖게 될거 같아서요
결국 실제로 가변인 것은 결국 어디선가 가변인 변수를 가질수 밖에 없다고 생각하는데 제가 잘못 이해한 것일까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceRecords.kt:
```diff
@@ -0,0 +1,15 @@
+        var max = 0
+        raceRecords.last().carRecords.forEach {
+            if (it.moveCount > max) max = it.moveCount
+        }
```
maxOf()를 사용해 볼까요?

- https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/max-of.html

Flamme1004K commented on src/main/kotlin/racingcar/domain/RandomNumberGenerator.kt:
```diff
@@ -0,0 +1,18 @@
+
+import java.security.SecureRandom
+
+class RandomNumberGenerator {
```
https://github.com/next-step/kotlin-racingcar/pull/1692#discussion_r1846408365

예전에 피드백 드린 부분이 있어요! 코틀린 클래스의 컨벤션에 대해 정리해 보시겠어요?

Flamme1004K commented on src/main/kotlin/racingcar/view/OutputView.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    private fun drawCar(carRecord: CarRecord) {
+        val sb = StringBuilder()
+        sb.append(carRecord.name)
+        sb.append(" : ")
+        repeat(carRecord.moveCount) {
+            sb.append("-")
+        }
+        println(sb.toString())
```
kotlin에서 StringBuilder와 "a" + "b"의 차이는 무엇일까요? 성능적으로도 차이가 있을까요?

gmltkddl1 replied:
+로 문자열을 추가하면 
String a
String b
이렇게 있을때 a + b를 실행하면 내부적으로 
String c가 생기고 여기에 a+b를 할당하는 방식으로 동작하는 것으로 알고있고
a,b가 메모리적으로 손해를 주는것으로 알고있습니다.

Flamme1004K replied:
JDK 9 에서의 a+b 는 어떻게 동작할까요?

gmltkddl1 replied:
+를 사용해도 쓰레기 남지 않는다고 하네요 이후부터 가독성을 고려해서 나은것을 사용해야겠네요

Flamme1004K commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -2,24 +2,21 @@ package racingcar
 import org.junit.jupiter.api.Test
+import racingcar.domain.Race
 
 class RaceTest {
```
많은 도메인에 대한 클래스가 생김에 따라서, 도메인 로직이 많이 나뉘어져 있어요. 새로운 도메인 클래스에 대한 테스트 케이스를 만들어 볼까요?

Flamme1004K reviewed (changes requested):
안녕하세요!

5단계 피드백 반영 잘 해주셨어요! 코멘트 남겼으니, 반영 후 다시 리뷰 요청 부탁드려요!

화이팅입니다!

gmltkddl1 committed:
refactor: step5 피드백 반영

Flamme1004K commented on src/test/kotlin/racingcar/RaceGameTest.kt:
```diff
@@ -0,0 +1,16 @@
+
+class RaceGameTest {
+    @Test
+    fun `가장 이동수가 큰차가 우승자이다`() {
```
피드백 반영 잘 해주셨어요! 우승자가 한 명인 테스트 좋습니다! 우승자가 여러 명인 테스트도 추가해 볼까요?

Flamme1004K commented on src/test/kotlin/racingcar/RaceGameTest.kt:
```diff
@@ -0,0 +1,16 @@
+import racingcar.domain.Race
+import racingcar.domain.RaceGame
+
+class RaceGameTest {
```
CarRecord, Race, RaceRecord, RaceRecords에 대한 테스트도 작성해 보시겠어요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceGame.kt:
```diff
@@ -2,12 +2,18 @@ package racingcar.domain
 
 class RaceGame(val race: Race) {
```
Race와 RaceGame 의 각 클래스는 어떠한 역할과 책임을 가지고 있는 클래스일까요?

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceGame.kt:
```diff
@@ -0,0 +1,19 @@
+package racingcar.domain
+
+class RaceGame(val race: Race) {
+    val raceRecords = RaceRecords()
```
raceRecords를 raceResult로 변경하고, getWinner() 함수를 raceResult로 옮기는 건 어떨까요?

raceResult로 변경하면, 

- 결과에 대한 기록을 가지고 온다.
- 결과에 대한 우승자를 가지고 온다.

라는 부분이 명확해질 듯해요.

Flamme1004K reviewed (changes requested):
안녕하세요! 

피드백 반영 매우 잘 해주셨습니다. 코멘트 남겼으니, 반영 후 리뷰 요청 부탁드립니다.

좀 만 더 힘내세요!

화이팅입니다!

gmltkddl1 committed:
refactor: step5 리팩토링

Car에서 NumberGenerator라는 인터페이스를 가지고 직접 번호를 생성하도록 수정

Flamme1004K commented on src/main/kotlin/racingcar/domain/Race.kt:
```diff
@@ -1,12 +1,20 @@
-    fun doRace(numbers: List<Int>) {
-        require(cars.size == numbers.size) { "랜덤 숫자가 차의 갯수보다 적게 들어왔어요" }
-        repeat(cars.size) { index -> cars[index].move(numbers[index]) }
+    companion object {
+        fun of(
+            names: List<String>,
+            numberGenerator: NumberGenerator,
+        ): Race {
+            val cars = names.map { Car(name = it, numberGenerator = numberGenerator) }
+            return Race(cars, numberGenerator)
+        }
```
피드백 반영 잘 해주셨어요!

Flamme1004K commented on src/main/kotlin/racingcar/domain/RaceGame.kt:
```diff
@@ -1,19 +1,25 @@
+    val race: Race,
+    var raceResult: RaceResult,
+) {
+    fun doRaceGame(numberOfTry: Int) {
+        repeat(numberOfTry) {
+            race.doRace()
+            val raceRecord = RaceRecord(race)
+            raceResult = raceResult.addRaceRecord(raceRecord)
```
> 제가 이해가 잘 안되는데 위 리스트는 실제로 추가가 필요한 리스트이고 피드백 주신대로 map으로 새로운 리스트를 만들게 되면
> var raceRecords: List = listOf() 이런식이 되야 할거 같습니다.
> 하지만 var 역시 가변이라고 생각되거든요
> val raceRecords: MutableList = mutableListOf() 이 구조보다 위와 같은 구조가
> 더 좋은 것 일까요?
> 아니면 그때 그때 아예 raceRecords를 새로 생성해는데 그렇게해도 이 객체를 프로퍼티로 가지는 객체가 가변인자를 갖게 될거 같아서요
> 결국 실제로 가변인 것은 결국 어디선가 가변인 변수를 가질수 밖에 없다고 생각하는데 제가 잘못 이해한 것일까요?

실제로 가변인 것은 결국 어디선가 가변인 변수를 가질 수 밖에 없다고 생각을 해요. raceRecords을 가변으로 과연 만들어야할까? 라는 부분도 고민을 해보시길 바래요. raceRecords는 그저 기록에 대해 보여주는 view 역할만 합니다. view 역할만 하기에 raceRecords는 과연 가변으로 내부에서도 외부에서도 값을 계속 변경하도록 노출시켜줘야하는가? 에 대한 부분에 대해서는 차라리 불변으로 지키는게 더 나아보인다는 의견을 드리고 싶네요.

Flamme1004K commented on src/test/kotlin/racingcar/CarTest.kt:
```diff
@@ -5,44 +5,69 @@ import org.assertj.core.api.Assertions.assertThat
+                }
+            }
+        val car = Car(name = "car1", numberGenerator = numberGenerator)
+        car.move()
+        car.move()
+        car.move()
         assertThat(car.moveCount).isEqualTo(3)
     }
 
     @Test
     fun notMoveTest() {
-        val car = Car(0, "car1")
-        car.move(2)
+        val numberGenerator =
+            object : NumberGenerator {
+                override fun getNumber(): Int {
+                    return 2
+                }
+            }
+        val car = Car(name = "car1", numberGenerator = numberGenerator)
+        car.move()
         assertThat(car.moveCount).isEqualTo(0)
     }
 
     @Test
     fun notMoveThreeTimesTest() {
-        val car = Car(0, "car1")
-        car.move(2)
-        car.move(1)
-        car.move(3)
+        val numberGenerator =
+            object : NumberGenerator {
+                override fun getNumber(): Int {
+                    return 2
+                }
+            }
+        val car = Car(name = "car1", numberGenerator = numberGenerator)
+        car.move()
+        car.move()
+        car.move()
```
나중에 이런 부분은 `@ParameterizedTest`의 `@MethodSource` 테스트로 하나의 테스트에서 여러 테스트를 할 수 있도록 만들어주면 좋겠네요~

Flamme1004K commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -2,21 +2,27 @@ package racingcar
-        assertThat(race.cars.map { it.moveCount }).containsExactly(0, 0, 0)
-        assertThat(race.cars.map { it.name }).containsExactly("car1", "car2", "car3")
-    }
+    private val numberGenerator =
+        object : NumberGenerator {
+            var flag = 3
+
+            override fun getNumber(): Int {
+                val result = flag
+                flag++
+                flag %= 10
+                return result
+            }
+        }
```
이런 경우에는 Test용 Fixture를 만들어서 따로 관리하면 좋겠네요~

Flamme1004K reviewed (approved):
안녕하세요~ 

그동안 5단계 미션 고생하셨어요!

이만 머지하도록 하겠습니다.

이후 미션도 화이팅입니다!

gmltkddl1 commented:
> 안녕하세요~
> 
> 그동안 5단계 미션 고생하셨어요!
> 
> 이만 머지하도록 하겠습니다.
> 
> 이후 미션도 화이팅입니다!

자세한 리뷰와 친절한 답변 감사했습니다.
